barman/postgres.py

```diff
@@ -1720,3 +1726,155 @@ def name_map(self):
             server_version = None
 
         return function_name_map(server_version)
+
+
+class StandbyPostgreSQLConnection(PostgreSQLConnection):
+    """
+    A specialised PostgreSQLConnection for standby servers.
+
+    Works almost exactly like a regular PostgreSQLConnection except it requires a
+    primary_conninfo option at creation time which is used to create a connection
+    to the primary for the purposes of forcing a WAL switch during the stop backup
+    process.
+
+    This increases the likelihood that backups against standbys with
+    `archive_mode = always` and low traffic on the primary are able to complete.
+    """
+
+    def __init__(
+        self,
+        conninfo,
+        primary_conninfo,
+        immediate_checkpoint=False,
+        slot_name=None,
+        application_name="barman",
+    ):
+        """
+        Standby PostgreSQL connection constructor.
+
+        :param str conninfo: Connection information (aka DSN) for the standby.
+        :param str primary_conninfo: Connection information for the primary.
+        :param bool immediate_checkpoint: Whether to do an immediate checkpoint
+            when a backup is started.
+        :param str|None slot_name: Replication slot name.
+        :param str: The application_name to use for this connection.
+        """
+        super(StandbyPostgreSQLConnection, self).__init__(
+            conninfo,
+            immediate_checkpoint=immediate_checkpoint,
+            slot_name=slot_name,
+            application_name=application_name,
+        )
+        self.primary_conninfo = primary_conninfo
+        # The standby connection has its own connection object used to talk to the
+        # primary when switching WALs.
+        self.primary = PostgreSQLConnection(
```

Comment:
Maybe Instanciate the `PostgreSQLConection instance` in server.Server._init_postgres() and give this instance when initiating `StandbyPostgreSQLConnection ` at the same place. This way both objects are created at the same place and less coupled.

Reply:
What I had intended to achieve here was making the `primary` reference an implementation detail of `StandbyPostgreSQLConnection` - the idea being that nothing which interacts with `StandbyPostgreSQLConnection` should have to care that it is using another PostgreSQL connection to the primary. However given this isn't actually the case (Server is already referencing `primary` directly) it would be better to create is in Server and pass it in.